pkg/querier/querier.go

```diff
@@ -921,3 +916,267 @@ func (q *SingleTenantQuerier) DetectedLabels(_ context.Context, _ *logproto.Dete
+		if err != nil {
+			return nil, err
+		}
+
+		iters = append(iters, ingesterIters...)
+	}
+
+	if !q.cfg.QueryIngesterOnly && storeQueryInterval != nil {
+		params := logql.SelectLogParams{
+			QueryRequest: &logproto.QueryRequest{
+				Selector:  req.Query,
+				Limit:     req.LineLimit,
+				Start:     storeQueryInterval.start,
+				End:       storeQueryInterval.end,
+				Direction: logproto.FORWARD,
+				Plan: &plan.QueryPlan{
+					AST: parsed,
+				},
+			},
+		}
+		level.Debug(spanlogger.FromContext(ctx)).Log(
+			"detected_fields", "true",
+			"msg", "querying store for detected fields",
+			"params", params)
+		storeIter, err := q.store.SelectLogs(ctx, params)
+		if err != nil {
+			return nil, err
+		}
+
+		iters = append(iters, storeIter)
+	}
+
+	finalIters := iter.NewMergeEntryIterator(ctx, iters, logproto.FORWARD)
+
+	level.Debug(spanlogger.FromContext(ctx)).Log(
+		"detected_fields", "true",
+		"have_iterators", finalIters.IsEmpty())
+
+	//TODO(twhitney): converting from a step to a duration should be abstracted and reused,
+	// doing this in a few places now.
+	streams, err := streamsForFieldDetection(finalIters, req.LineLimit, time.Duration(req.Step*1e6))
+	if err != nil {
+		return nil, err
+	}
+
+	level.Debug(spanlogger.FromContext(ctx)).Log(
+		"detected_fields", "true",
+		"msg", fmt.Sprintf("found %d streams to detect fields for", len(streams)))
+	detectedFields := parseDetectedFields(ctx, req.FieldLimit, streams)
+
+	fields := make([]*logproto.DetectedField, len(detectedFields))
+	fieldCount := 0
+	for k, v := range detectedFields {
+		fields[fieldCount] = &logproto.DetectedField{
+			Label:       k,
+			Type:        v.fieldType,
+			Cardinality: v.Estimate(),
+		}
+
+		fieldCount++
+	}
+
+	return &logproto.DetectedFieldsResponse{
+		Fields: fields,
+	}, nil
+}
+
+type parsedFields struct {
+	sketch         *hyperloglog.Sketch
+	isTypeDetected bool
+	fieldType      logproto.DetectedFieldType
+}
+
+func newParsedFields() *parsedFields {
+	return &parsedFields{
+		sketch:         hyperloglog.New(),
+		isTypeDetected: false,
+		fieldType:      logproto.DetectedFieldString,
+	}
+}
+
+func (p *parsedFields) Insert(value string) {
+	p.sketch.Insert([]byte(value))
+}
+
+func (p *parsedFields) Estimate() uint64 {
+	return p.sketch.Estimate()
+}
+
+func (p *parsedFields) DetermineType(value string) {
+	p.fieldType = determineType(value)
+	p.isTypeDetected = true
+}
+
+func determineType(value string) logproto.DetectedFieldType {
+	if _, err := strconv.ParseInt(value, 10, 64); err == nil {
+		return logproto.DetectedFieldInt
+	}
+
+	if _, err := strconv.ParseFloat(value, 64); err == nil {
+		return logproto.DetectedFieldFloat
+	}
+
+	if _, err := strconv.ParseBool(value); err == nil {
+		return logproto.DetectedFieldBoolean
+	}
+
+	if _, err := time.ParseDuration(value); err == nil {
+		return logproto.DetectedFieldDuration
+	}
+
+	if _, err := humanize.ParseBytes(value); err == nil {
```

Comment:
this could be a string.Contains like we do for the level label detection. I suspect a small benchmark with some special log line will help. Might also be true for duration.